File: 611-valid-triangle-number-2.java
import java.util.Arrays;

class ValidTriangleNumberSolution {
    /**
     * For a valid triangle, sum of 2 sides must be larger than 3rd side,
     * i.e. a + b > c
     * We can first sort the nums list,
     * then we have a loop that goes through each number,
     * setting that as `c`.
     * Next, we use 2 pointers to find all valid pairs of `a` and `b` that is
     * greater than `c`.
     */

    public int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        int result = 0;

        for (int c = 0; c < nums.length; c++) {
            int a = 0, b = c - 1;
            while (a < b) {
                if (nums[a] + nums[b] <= nums[c]) {
                    a++;
                } else {
                    // if the sum is greater, then all pairs from a, a+1... to b will work, so we
                    // need to get all triangles between b and a
                    result += b - a;
                    b--;
                }
            }
        }

        return result;
    }

    public static void main(String[] args) {
        ValidTriangleNumberSolution sol = new ValidTriangleNumberSolution();
        System.out.println("Output: " + sol.triangleNumber(new int[] { 2, 2, 3, 4 }) + ", expected: 3");
        System.out.println("Output: " + sol.triangleNumber(new int[] { 4, 2, 3, 4 }) + ", expected: 4");
    }

}

File: 611-valid-triangle-number-2.py
class Solution:
    def triangleNumber(self, nums: list[int]) -> int:
        '''
        For a valid triangle, sum of 2 sides must be larger than 3rd side, i.e. a + b > c
        We can first sort the nums list, then we have a loop that goes through each number, setting that as `c`.
        Next, we use 2 pointers to find all valid pairs of `a` and `b` that is greater than `c`.
        '''

        nums.sort()
        result = 0
        n = len(nums)
        for c in range(n):
            a = 0
            b = c - 1
            while a < b:
                sum = nums[a] + nums[b]
                if sum <= nums[c]:
                    a += 1
                else:
                    # if the sum is greater, then all pairs from a, a+1... to b will work, so we need to get all triangles between b and a
                    result += b - a
                    b -= 1

        return result


sol = Solution()
print(f'output: {sol.triangleNumber([2, 2, 3, 4])}, expected: 3')
print(f'output: {sol.triangleNumber([4, 2, 3, 4])}, expected: 4')
